Sources/Authentication/Passwords/PBKDF2Hasher.swift

```
@@ -0,0 +1,180 @@
+            outputByteCount: parsed.hash.count,
+            unsafeUncheckedRounds: parsed.iterations
+        )
+
+        let keyData = unsafe key.withUnsafeBytes { unsafe Data($0) }
+
+        return keyData.elementsEqual(parsed.hash)
+    }
+
+    private struct ParsedPassword {
+        let algorithm: KDF.Insecure.PBKDF2.HashFunction
+        let iterations: Int
+        let salt: [UInt8]
+        let hash: [UInt8]
+    }
+
+    private static func parsePassword(_ string: String) -> ParsedPassword? {
+        // Expected format: $pbkdf2-<alg>$<iterations>$<b64salt>$<b64hash>
+        let parts = string.split(separator: "$", omittingEmptySubsequences: true)
+        guard parts.count == 4 else { return nil }
+
+        // Parse algorithm
+        let algPart = String(parts[0])
+        guard
+            algPart.hasPrefix("pbkdf2-"),
+            let algorithm = hashFunction(from: String(algPart.dropFirst(7)))
+        else {
+            return nil
+        }
+
+        // Parse iterations
+        guard let iterations = Int(parts[1]) else {
+            return nil
+        }
+
+        // Parse salt
+        guard let saltData = Data(base64Encoded: String(parts[2])) else {
+            return nil
+        }
+
+        // Parse hash
+        guard let hashData = Data(base64Encoded: String(parts[3])) else {
+            return nil
+        }
+
+        return ParsedPassword(
+            algorithm: algorithm,
+            iterations: iterations,
+            salt: Array(saltData),
+            hash: Array(hashData)
+        )
+    }
+
+    private static func algorithmIdentifier(for hashFunction: KDF.Insecure.PBKDF2.HashFunction) -> String {
+        switch hashFunction {
+        case .sha256: "sha256"
+        case .sha384: "sha384"
+        case .sha512: "sha512"
+        case .insecureSHA1: "sha1"
+        case .insecureSHA224: "sha224"
+        case .insecureMD5: "md5"
+        default: fatalError("Unsupported hash function")
```

I'd much rather rather throw an error here. It's fine now but as it gets more complicated, if other algorithms get added and this gets missed then there's a future crash (we've hit issues like this historically so while we can, lets make it throwing)

---

Actually in hindsight, this also means we need to expose CryptoExtras as our public interface. Lets create our own enum, (we can use new extensible enums if we want), then it can become internal import